perspective_reddit_bot/moderate_subreddit.py

```diff
@@ -0,0 +1,152 @@
+  mods = subreddit.moderator()
+  return reddit.user.me() in mods
+
+
+def score_subreddit(creds_dict,
+                    subreddit_name,
+                    rules,
+                    models,
+                    output_path=None):
+  """Scores a stream of commments via the perspective API and applies pre-defined moderation rules.
+
+  Args:
+    creds_dict: (dict) A dictionary of API credentials for Perspective and Reddit.
+    subreddit_name: (str) The name of the subreddit to stream.
+    rules: (list) A list of rules to apply to each comment.
+    models: (list) A list of models that the API must call to apply rules.
+    output_path: (str, optional) If supplied, all comments and scores will be written to this path.
+  """
+
+  reddit = praw.Reddit(client_id=creds_dict['reddit_client_id'],
+                       client_secret=creds_dict['reddit_client_secret'],
+                       user_agent=creds_dict['reddit_user_agent'],
+                       username=creds_dict['reddit_username'],
+                       password=creds_dict['reddit_password'])
+  subreddit = reddit.subreddit(subreddit_name)
+
+  try:
+    mod_permissions = bot_is_mod(reddit, subreddit)
+  except Exception:
+    mod_permissions = False
+  if mod_permissions:
+    print('Bot is moderator of subreddit.')
+    print('Moderation actions will be applied.')
+  else:
+    print('Bot is not moderator of subreddit.')
+    print('Moderation actions will not be applied.')
+
+  perspective = perspective_client.PerspectiveClient(creds_dict['perspective_api_key'])
+
+  for i, comment in enumerate(subreddit.stream.comments()):
+    try:
+      if i%100 == 0 and i > 0:
+        print(i)
+      if len(comment.body) > 20000:
+        print('Comment too long, skipping...')
+        continue
+      df = pd.DataFrame()
+      df = df.append({TEXT_COLUMN: comment.body}, ignore_index=True)
+      scored_df = score_dataset.score_dataframe(df,
+                                                TEXT_COLUMN,
+                                                models,
+                                                perspective,
+                                                language=LANGUAGE)
+      if output_path:
+        with open(output_path, 'a') as f:
+          scored_df.to_json(f, orient='records', lines=True)
+
+      for rule in rules:
+        if rule.check_model_rules(scored_df):
+          print('----------')
+          print('Comment #%s: ' % i)
+          print(comment.body.encode('utf-8'))
+          print('Rule: %s' % rule)
+          print('Action: %s' % rule.action_name)
+          if subreddit_name in ['all', 'popular']:
+            print('Subreddit: %s' % comment.subreddit)
+
+          if mod_permissions:
+            rule.apply_action(comment)
+            print('ACTION APPLIED')
+          else:
+            print('ACTION NOT APPLIED')
+
+          print('----------')
+    except Exception as e:
+      print('Skipping comment due to exception: %s' % str(e))
```

Comment:
i think the `%s` formatter means you don't need to call `str()` explicitly

Reply:
Good point! Done.